### PR TITLE
Implement frontend for test execution review

### PR DESCRIPTION
**Pull request by andrejvelichkovski**
This PR resolves RTW-204 and implements the frontend functionality for test execution review.

# Implementation Changes

As part of this change, a few things have been updated:
* New popover for the review process was added
* Button to trigger the popover
* Riverpod provider ArtefactBuilds was updated to also include functionality for patching the test execution

# Screenshots
Screenshot from 2024-01-03 09-17-27
Screenshot from 2024-01-03 09-17-34
Screenshot from 2024-01-03 09-17-48

# Design changes

I have changed the design slightly, as I couldn't find a way to use `YaruPopupMenuButton` which is specified in the design to be a standard button like TextButton.

# Unresolved issues

When updating an existing Test Execution Review, I managed to load the selected decisions. However, I couldn't find a way to load the review comment into the `TextField` widget.

### CODE REVIEW

File: frontend/lib/ui/artefact_dialog/test_execution_review.dart [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+              hintText: 'Insert review comment',
+            ),
+            onChanged: (value) {
+              setState(() {
+                reviewComment = value;
+              });
+            },
+          ),
+          Padding(
+            padding: const EdgeInsets.only(top: 8.0),
```
Comment (omar-selo) [inline frontend/lib/ui/artefact_dialog/test_execution_review.dart:172]: typically spacing between widgets is done using `SizedBox` rather than padding

**Review by omar-selo — CHANGES_REQUESTED** [review]
Thanks for the changes. Most are small points but one issue is about passing `artefactId`

File: frontend/lib/providers/artefact_builds.dart [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
+    await dio.patch(
+      '/v1/test-executions/$testExecutionId',
+      data: {
+        'review_decision': reviewDecision.map((e) => e.toJson()).toList(),
+        'review_comment': reviewComment,
+      },
+    );
+
+    final updatedState = await build(artefactId);
+    state = AsyncData(updatedState);
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/providers/artefact_builds.dart:38]: I was also thinking about this, but the issue with the first option was that the PATCH endpoint modified the TestExecution object, not the ArtefactBuild object, and the riverpod provider is for the ArtefactBuilds. 

I guess one option would be to have this endpoint return the whole ArtefactBuild object, which is a bit counter-intuitive as we are sending PATCH request to a TestExecution object.

File: frontend/lib/ui/artefact_dialog/test_execution_review.dart [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  bool _enableApprovedOptions() {
+    return !(reviewDecisions[TestExecutionReviewDecision.rejected] ?? false);
+  }
+
+  bool _enableRejectedOptions() {
+    return !TestExecutionReviewDecision.values.any(
+      (element) => (element != TestExecutionReviewDecision.rejected &&
+          (reviewDecisions[element] ?? false)),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/ui/artefact_dialog/test_execution_review.dart:96]: I'm not sure if you mean getter on the TestExecutionReviewDecision enum. In that case, I'm not sure how we can do it, as these are stateful (they depend on the list of selected decisions).

File: frontend/lib/providers/artefact_builds.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (omar-selo) [reply frontend/lib/providers/artefact_builds.dart:38]: True, I hadn't taken that into consideration. Then I guess options 2 or 3 would work. Where for 3 we would just replace the child test execution object in the artefact build

File: frontend/lib/ui/artefact_dialog/test_execution_review.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (omar-selo) [reply frontend/lib/ui/artefact_dialog/test_execution_review.dart:96]: Oh I wasn't very clear. I mean class getters https://dart.dev/language/methods#getters-and-setters

[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/ui/artefact_dialog/test_execution_review.dart:96]: Thanks! I didn't know about getters/setters, but will update the code.

File: frontend/lib/ui/artefact_dialog/artefact_build_expandable.dart [hunk 4]
```diff
@@ -7,9 +7,14 @@ import '../spacing.dart';
 import 'test_execution_expandable.dart';
 
 class ArtefactBuildExpandable extends StatelessWidget {
-  const ArtefactBuildExpandable({super.key, required this.artefactBuild});
+  const ArtefactBuildExpandable({
+    super.key,
+    required this.artefactBuild,
+    required this.artefactId,
+  });
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/ui/artefact_dialog/artefact_build_expandable.dart:14]: I agree passing the prop is not perfect, but I wasn't sure how to do it otherwise.

We don't need the artefactId to send this request, only the test execution id, but we need the artefactID to get the ArtefactBuild object from riverpod.

I guess the best approach would be to have TestExecutionProvier which would not send requests to the backend but rather take the data from the ArtefactBuildsProvider, similarly to how we have the StageArtefactsProvider take the data from FamilyArtefactsProvider.

However, my only concern is how we would update the ArtefactBuildProvider state in that case, as it might include some incosistent data with the TestExecutionProvider, in case we change something. I guess in that case we need to make sure to never use the .testExecution object from ArtefactBuildProvider, but to always use the TestExecutionProvider, which makes sense.

I'm not sure about the flutter data package, as I don't have a strong understanding of how Flutter works yet. I will try to read more around it.

How do you suggest we go about this?

File: frontend/lib/ui/artefact_dialog/test_execution_review.dart [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      onTap: () {
+        showPopover(
+          context: context,
+          bodyBuilder: (context) => TestExecutionPopOver(
+            testExecution: testExecution,
+            artefactId: artefactId,
+          ),
+          direction: PopoverDirection.bottom,
+          width: 500,
+          height: 500,
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/ui/artefact_dialog/test_execution_review.dart:53]: Yes, the popover becomes scrollable

Screenshot from 2024-01-03 12-21-30

File: frontend/pubspec.yaml [hunk 6]
```diff
@@ -15,6 +15,7 @@ dependencies:
   go_router: ^11.1.4
   intersperse: ^2.0.0
   json_annotation: ^4.8.1
+  popover: 0.2.8
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/pubspec.yaml]: It kept downloading version 0.2.9 for me locally, which wasn't working with the Git checker. So I fixed it to 0.2.8. I will check the automated checks to see if they fail again

File: frontend/lib/providers/artefact_builds.dart [hunk 7]
```diff
@@ -1,20 +1,40 @@
...
+    String reviewComment,
+    List<TestExecutionReviewDecision> reviewDecision,
+  ) async {
+    final dio = ref.watch(dioProvider);
+
+    await dio.patch(
+      '/v1/test-executions/$testExecutionId',
+      data: {
+        'review_decision': reviewDecision.map((e) => e.toJson()).toList(),
+        'review_comment': reviewComment,
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/providers/artefact_builds.dart]: I have refactored to have these value as a constant and generate inside the class

File: frontend/lib/providers/artefact_builds.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply frontend/lib/providers/artefact_builds.dart]: I was thinking it would be cleaner to go with the second approach, although it involves an extra network request to rebuild the state, but it makes sure we always put the correct data in the state.

File: frontend/lib/providers/artefact_builds.dart [hunk 8]
```diff
@@ -19,22 +19,47 @@ class ArtefactBuilds extends _$ArtefactBuilds {
...
+            artefactBuild.testExecutions.map(
+              (e) {
+                if (e.id == testExecutionId) {
+                  return TestExecution.fromJson(responseData);
+                }
+                return e;
+              },
+            ),
+          ),
+        ),
```
Comment (omar-selo) [inline frontend/lib/providers/artefact_builds.dart:42]: I don't think that this code is completely correct because at doesn't make copies of other test executions and doesn't copy environments. And state [should be immutable](https://riverpod.dev/docs/concepts/why_immutability#using-immutable-state). I recommend using `copyWith` for artefact builds and test executions and possibly adding a helper function.

File: frontend/lib/ui/artefact_dialog/test_execution_pop_over.dart [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                      ? (bool? value) {
+                          setState(() {
+                            if (reviewDecisions.contains(e)) {
+                              reviewDecisions.remove(e);
+                            } else {
+                              reviewDecisions.add(e);
+                            }
+                          });
+                        }
+                      : null,
```
Comment (omar-selo) [inline frontend/lib/ui/artefact_dialog/test_execution_pop_over.dart:77]: can you pull `onChanged` as a separate helper method/function as it's a bit hard to follow this way

File: frontend/lib/ui/artefact_dialog/test_execution_review.dart [hunk 10]
```diff
@@ -40,6 +37,7 @@ class TestExecutionReviewButton extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
+    final artefactId = AppRoutes.artefactIdFromContext(context);
```
Comment (omar-selo) [inline frontend/lib/ui/artefact_dialog/test_execution_review.dart:40]: given that this widget doesn't use `artefactId` perhaps `TestExecutionPopOver` should be the one to get it from context?

**Review by omar-selo — CHANGES_REQUESTED** [review]
Looks much better, just a few more comments

File: frontend/lib/ui/artefact_dialog/test_execution_review.dart (hunk 10 above)
Comment (andrejvelichkovski) [reply frontend/lib/ui/artefact_dialog/test_execution_review.dart:40]: Actually, I tried this initially and it didn't work, I think it is because the Popover is a separate layer in the render. After some searching, I found out about this [open issue in Flutter](https://github.com/flutter/flutter/issues/131154) which seems to be going resolved in a [separate issue](https://github.com/flutter/flutter/issues/129833) but doesn't seem to be closed yet.

File: backend/test_observer/data_access/models.py [hunk 11]
```diff
@@ -166,7 +166,9 @@ class ArtefactBuild(Base):
         back_populates="builds", foreign_keys=[artefact_id]
     )
     test_executions: Mapped[list["TestExecution"]] = relationship(
-        back_populates="artefact_build", cascade="all, delete"
+        back_populates="artefact_build",
+        cascade="all, delete",
+        order_by="TestExecution.environment_id",
```
[the remark this answers is not in the document]
Comment (andrejvelichkovski) [reply backend/test_observer/data_access/models.py:171]: Good point, I have updated to order it at the controller level

File: frontend/lib/providers/artefact_builds.dart (hunk 8 above)
Comment (andrejvelichkovski) [reply frontend/lib/providers/artefact_builds.dart]: I see, this code is indeed not really good idea, but I think using copyWith wont be much simpler, as we again need to rebuild at least one artefact build. It seems like a better option is to make that additional network request after all..

Comment (omar-selo) [reply frontend/lib/providers/artefact_builds.dart]: adding a network request feels like choosing a worst solution for the sake of better code. We don't need to rebuild one artefact build. Once we copyWith everything, we just modify one artefact build

Comment (omar-selo) [reply frontend/lib/providers/artefact_builds.dart]: Oh ignore my point about modifying one artefact build as it's a freezed object so it can't be modified

Comment (andrejvelichkovski) [reply frontend/lib/providers/artefact_builds.dart]: I agree with your points, I think my implementation should be updating the state correctly now.

File: frontend/lib/ui/artefact_dialog/test_execution_pop_over.dart (hunk 9 above)
Comment (andrejvelichkovski) [reply frontend/lib/ui/artefact_dialog/test_execution_pop_over.dart]: I now extracted this in a separate function.

File: frontend/lib/ui/artefact_dialog/artefact_dialog.dart [hunk 12]
```diff
@@ -16,47 +17,59 @@ class ArtefactDialog extends ConsumerWidget {
...
+                        crossAxisAlignment: CrossAxisAlignment.start,
+                        children: [
+                          ArtefactDialogHeader(artefact: artefact),
+                          const SizedBox(height: Spacing.level4),
+                          ArtefactDialogInfoSection(artefact: artefact),
+                          const SizedBox(height: Spacing.level4),
+                          Expanded(
+                            child: ArtefactDialogBody(artefact: artefact),
+                          ),
+                        ],
```
Comment (andrejvelichkovski) [inline frontend/lib/ui/artefact_dialog/artefact_dialog.dart:63]: This is the change to fix the async error.

**Comment by andrejvelichkovski** [comment]
> It looks like the c3 link is still there in the code, but I don't see it in the screenshots you posted. Is that just something off with the screenshot, or was there some UI change that somehow removed the c3 link here?

The C3 Link button is conditionally rendered, it just happened that none of these test executions have it set up in the database, so it is not shown on the frontend either.

File: frontend/lib/ui/artefact_dialog/artefact_dialog.dart (hunk 12 above)
Comment (omar-selo) [reply frontend/lib/ui/artefact_dialog/artefact_dialog.dart:63]: Thanks for fixing my error
